File: app/src/test/java/io/github/tobyhs/weatherweight/AppModuleTest.java
package io.github.tobyhs.weatherweight;

import android.content.SharedPreferences;

import androidx.test.core.app.ApplicationProvider;

import com.github.tobyhs.rxsecretary.android.AndroidSchedulerProvider;

import com.google.gson.Gson;

import okhttp3.OkHttpClient;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import io.github.tobyhs.weatherweight.api.accuweather.AccuWeatherApiKeyInterceptor;
import io.github.tobyhs.weatherweight.api.accuweather.AccuWeatherService;
import io.github.tobyhs.weatherweight.data.AccuWeatherRepository;
import io.github.tobyhs.weatherweight.data.WeatherRepository;
import io.github.tobyhs.weatherweight.storage.LastForecastStore;
import io.github.tobyhs.weatherweight.storage.SharedPrefLastForecastStore;

import static org.hamcrest.CoreMatchers.hasItem;
import static org.hamcrest.CoreMatchers.instanceOf;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.isA;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.mock;

@RunWith(RobolectricTestRunner.class)
@Config(application = App.class)
public class AppModuleTest {
    private static final String TRUST_STORE_PROPERTY = "javax.net.ssl.trustStore";
    private String originalTrustStore;

    private App app;
    private AppModule module;

    @Before
    public void setup() {
        originalTrustStore = System.getProperty(TRUST_STORE_PROPERTY);
        System.setProperty(TRUST_STORE_PROPERTY, "NONE");

        app = ApplicationProvider.getApplicationContext();
        module = new AppModule(app);
    }

    @After
    public void teardown() {
        if (originalTrustStore != null) {
            System.setProperty(TRUST_STORE_PROPERTY, originalTrustStore);
        }
    }

    @Test
    public void provideApp() {
        assertThat(module.provideApp(), is(app));
    }

    @Test
    public void provideSharedPreferences() {
        assertThat(module.provideSharedPreferences(), is(notNullValue()));
    }

    @Test
    public void provideSchedulerProvider() {
        assertThat(module.provideSchedulerProvider(), is(instanceOf(AndroidSchedulerProvider.class)));
    }

    @Test
    public void provideGson() {
        assertThat(AppModule.provideGson(), is(notNullValue()));
    }

    @Test
    public void provideAccuWeatherOkHttp() {
        OkHttpClient client = module.provideAccuWeatherOkHttp();
        assertThat(client.interceptors(), hasItem(isA(AccuWeatherApiKeyInterceptor.class)));
        assertThat(client.cache(), is(notNullValue()));
    }

    @Test
    public void provideAccuWeatherRetrofit() {
        OkHttpClient client = mock(OkHttpClient.class);
        Retrofit retrofit = module.provideAccuWeatherRetrofit(AppModule.provideGson(), client);
        assertThat(retrofit.baseUrl().toString(), is("https://dataservice.accuweather.com/"));
        assertThat(retrofit.callFactory(), is(client));
        assertThat(retrofit.callAdapterFactories(), hasItem(isA(RxJava3CallAdapterFactory.class)));
        assertThat(retrofit.converterFactories(), hasItem(isA(GsonConverterFactory.class)));
    }

    @Test
    public void provideAccuWeatherService() {
        Retrofit retrofit = new Retrofit.Builder().baseUrl("http://localhost/").build();
        assertThat(module.provideAccuWeatherService(retrofit), is(notNullValue()));
    }

    @Test
    public void provideWeatherRepository() {
        AccuWeatherService service = mock(AccuWeatherService.class);
        WeatherRepository repo = module.provideWeatherRepository(service);
        assertThat(repo, is(instanceOf(AccuWeatherRepository.class)));
    }

    @Test
    public void provideLastForecastStore() {
        SharedPreferences sharedPreferences = mock(SharedPreferences.class);
        Gson gson = AppModule.provideGson();
        LastForecastStore store = module.provideLastForecastStore(sharedPreferences, gson);
        assertThat(store, is(instanceOf(SharedPrefLastForecastStore.class)));
    }
}

File: app/src/test/java/io/github/tobyhs/weatherweight/AppModuleTest.kt
package io.github.tobyhs.weatherweight

import android.annotation.SuppressLint
import androidx.test.core.app.ApplicationProvider

import com.github.tobyhs.rxsecretary.android.AndroidSchedulerProvider

import io.github.tobyhs.weatherweight.api.accuweather.AccuWeatherApiKeyInterceptor
import io.github.tobyhs.weatherweight.api.accuweather.AccuWeatherService
import io.github.tobyhs.weatherweight.data.AccuWeatherRepository
import io.github.tobyhs.weatherweight.storage.FileLastForecastStore

import java.time.LocalDate
import java.time.ZonedDateTime

import okhttp3.OkHttpClient

import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.hasItem
import org.hamcrest.CoreMatchers.instanceOf
import org.hamcrest.CoreMatchers.isA
import org.hamcrest.CoreMatchers.notNullValue
import org.hamcrest.MatcherAssert.assertThat

import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

import org.mockito.Mockito.mock

import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

import retrofit2.Retrofit
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory

@RunWith(RobolectricTestRunner::class)
@Config(application = App::class)
class AppModuleTest {
    private var originalTrustStore: String? = null
    private lateinit var app: App
    private val module = AppModule()

    @Before
    fun setup() {
        originalTrustStore = System.getProperty(TRUST_STORE_PROPERTY)
        System.setProperty(TRUST_STORE_PROPERTY, "NONE")
        app = ApplicationProvider.getApplicationContext()
    }

    @After
    fun teardown() {
        if (originalTrustStore != null) {
            System.setProperty(TRUST_STORE_PROPERTY, originalTrustStore)
        }
    }

    @Test
    fun provideSchedulerProvider() {
        assertThat(
            module.provideSchedulerProvider(),
            instanceOf(AndroidSchedulerProvider::class.java)
        )
    }

    @Test
    fun provideAccuWeatherOkHttp() {
        val client = module.provideAccuWeatherOkHttp(app)
        assertThat(client.interceptors(), hasItem(isA(AccuWeatherApiKeyInterceptor::class.java)))
        assertThat(client.cache(), notNullValue())
    }

    @Test
    fun provideAccuWeatherRetrofit() {
        val client = mock(OkHttpClient::class.java)
        val retrofit = module.provideAccuWeatherRetrofit(
            client,
            AppModule.provideMoshi(),
        )
        assertThat(retrofit.baseUrl().toString(), equalTo("https://dataservice.accuweather.com/"))
        assertThat(retrofit.callFactory(), equalTo(client))
        assertThat(
            retrofit.callAdapterFactories(),
            hasItem(isA(RxJava3CallAdapterFactory::class.java))
        )
        assertThat(retrofit.converterFactories(), hasItem(isA(MoshiConverterFactory::class.java)))
    }

    @Test
    fun provideAccuWeatherService() {
        val retrofit = Retrofit.Builder().baseUrl("http://localhost/").build()
        assertThat(module.provideAccuWeatherService(retrofit), notNullValue())
    }

    @Test
    fun provideWeatherRepository() {
        val service = mock(AccuWeatherService::class.java)
        val repo = module.provideWeatherRepository(service)
        assertThat(repo, instanceOf(AccuWeatherRepository::class.java))
    }

    @Test
    fun provideLastForecastStore() {
        val moshi = AppModule.provideMoshi()
        val store = module.provideLastForecastStore(app, moshi)
        assertThat(store, instanceOf(FileLastForecastStore::class.java))

        val fileLastForecastStore = store as FileLastForecastStore
        assertThat(fileLastForecastStore.directory, equalTo(app.cacheDir))
        assertThat(fileLastForecastStore.moshi, equalTo(moshi))
    }

    @Test
    @SuppressLint("CheckResult")
    fun provideMoshi() {
        val moshi = AppModule.provideMoshi()
        // Check that the following don't throw an IllegalArgumentException
        moshi.adapter(LocalDate::class.java)
        moshi.adapter(ZonedDateTime::class.java)
    }

    companion object {
        private const val TRUST_STORE_PROPERTY = "javax.net.ssl.trustStore"
    }
}
